comms/src/connection_manager/connections.rs

```diff
@@ -0,0 +1,382 @@
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::{
+    error::ConnectionManagerError,
+    repository::{ConnectionRepository, PeerConnectionEntry, Repository},
+    Result,
+};
+
+use crate::{
+    connection::{
+        self,
+        curve_keypair::{CurvePublicKey, CurveSecretKey},
+        net_address::ip::SocketAddress,
+        Context,
+        CurveEncryption,
+        Direction,
+        NetAddress,
+        NetAddresses,
+        PeerConnection,
+        PeerConnectionContextBuilder,
+        PeerConnectionState,
+    },
+    peer_manager::node_id::NodeId,
+    utils,
+};
+
+use crate::connection::InprocAddress;
+use std::{
+    collections::HashMap,
+    net::IpAddr,
+    ops::Range,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+    thread::JoinHandle,
+    time::Duration,
+};
+
+pub struct PeerConnectionConfig {
+    pub max_message_size: u64,
+    pub max_connect_retries: u16,
+    pub socks_proxy_address: Option<SocketAddress>,
+    pub consumer_address: InprocAddress,
+    pub port_range: Range<u16>,
+    pub host: IpAddr,
+    pub establish_timeout: Duration,
+}
+
+/// Indicates the type of connection to establish.
+pub enum ConnectionDirection {
+    /// Listen for a connection from a peer and possibly accept their connection
+    Inbound {
+        node_id: NodeId,
+        secret_key: CurveSecretKey,
+    },
+    /// Connect to a peer and wait for them to possibly accept
+    Outbound {
+        node_id: NodeId,
+        net_addresses: NetAddresses,
+        server_public_key: CurvePublicKey,
+    },
+}
+
+/// Stores, and establishes the live peer connections
+pub struct LivePeerConnections {
+    context: Context,
+    repository: RwLock<ConnectionRepository>,
+    pub(crate) config: PeerConnectionConfig,
+    connection_thread_handles: RwLock<HashMap<NodeId, JoinHandle<connection::Result<()>>>>,
+}
+
+impl LivePeerConnections {
+    /// Create a new live peer connection
+    pub fn new(context: Context, config: PeerConnectionConfig) -> Self {
+        Self {
+            context,
+            config,
+            repository: RwLock::new(ConnectionRepository::default()),
+            connection_thread_handles: RwLock::new(HashMap::new()),
```

Comment:
The `repository` and `connection_thread_handles` RwLocks might get out of sync, might be better to handle them using a single RwLock.